flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,462 @@
+
+The definition of the `CapabilityController`.
+
+```cadence
+struct CapabilityController {
+   // The block height when the capability was created.
+   let issueHeight: UInt64
+   // The Type of the capability, i.e.: the T in Capability<T>.
+   let borrowType: Type
+   // The id of the related capability.
+   // This is the UUID of the created capability.
+   // All copies of the same capability will have the same UUID
+   let capabilityID: UInt64
+   // Is the capability revoked.
+   let isRevoked: Bool
+
+   // Returns the targeted storage path of the capability.
+   fun target(): StoragePath
+   // Revoke the capability making it no longer usable.
+   // When borrowing from a revoked capability the borrow returns nil.
+   fun revoke()
+   // Retarget the capability.
+   // This moves the CapCon from one CapCon array to another.
+   // If the new target is not a valid target this will panic.
+   fun retarget(target: StoragePath)
+}
+```
+
+Capabilities would expose the new id field `let id: UInt64` in addition to the account field they already expose.
+
+The capability related methods in the `AuthAccount` and in the `PublicAccount` would be moved to a `capabilities` namespace, similar to how the contract methods are in the [contracts namespace](https://developers.flow.com/cadence/language/accounts).
+
+The `AuthAccount` would get new methods to create or get or iterate through capabilities and to get or iterate through CapCons in order to manage capabilities. Methods used for (un)linking would be removed as they are no longer needed.
+
+```cadence
+struct AuthAccount {
+   // ...
+   // removed: fun link<T: &Any>(_ newCapabilityPath: CapabilityPath, target: Path): Capability<T>?
+   // removed: fun getLinkTarget(_ path: CapabilityPath): Path?
+   // removed: fun unlink(_ path: CapabilityPath)
+   // moved & renamed: fun getCapability<T>(_ path: CapabilityPath): Capability<T>
+   // moved & renamed: fun forEachPublic(_ function: ((PublicPath, Type): Bool))
+   let capabilities: AuthAccount.Capabilities
+   struct Capabilities {
+      // get returns the capability at the public path, if one was stored there.
+      fun get<T>(_ path: PublicPath): Capability<T>?
+      // borrow is equivalent to `get(path)!.borrow()` if `get(path)` exists `nil` otherwise
+      fun borrow<T>(_ path: PublicPath): T?
+      // For each iterates through all the public capabilities of the public account.
+      // If function returns false, the iteration ends.
+      fun forEach(_ function: ((PublicPath, Type): Bool))
+
+      // Get capability controller for capability with the specified id
+      // If the id does not reference an existing capability
+      // or the capability does not target a storage path on this address, return nil
+      fun getController(byCapabilityID: UInt64): CapabilityController?
```

Comment:
Given `CapabilityController` is a value, it can be copied. Would that be an issue to have multiple CapCons to a single Capability?
For e.g: Say `capConCopy` is a copy of `capCon`, and if someone calls `capCon.revoke()`, then it would also need to update the `capConCopy.isRevoked` field as well.

So, wouldn't it be better to return a reference to the `CapabilityController` instead of the value itself?

Reply:
See https://github.com/onflow/flow/pull/798#issuecomment-1324705992 and https://github.com/onflow/flow/pull/798#issuecomment-1325283451. 

Though I agree with you and @bluesign that cap cons are references, I still feel like it would be nice to keep the API simple, and don't "burden" users with this detail. We do this already today for e.g. `AuthAccount`.

Reply:
Sorry, this wen't completely out of my radar.

> I still feel like it would be nice to keep the API simple, and don't "burden" users with this detail.

How would a reference make the API complex? Let's say we return a reference as a CapCon, then users will only use it to access fields, and call functions on it. So the code they would write is exactly the same as using a non-reference struct value. Is there anything else they would want to do with the CapCon like storing/moving etc? (pretty much agree with @bluesign on https://github.com/onflow/flow/pull/798#issuecomment-1325342206)

As a dev, it would make me feel a bit confused if some structs behave as value types while some behave like references. Yes, we may have done it for AuthAccount but I think it would be bad to spread that behavior to other types as well (and going one step further, maybe also fix it in AuthAccount! But it'll be hard to undo)

Reply:
This concern has been addressed, see below: It was decided to return references